source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+collection in the specified bucket and uses the information to reconstruct
+the file from documents in the ``chunks`` collection. You can read the file
+into memory or output it to a stream.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To use GridFS, first create a bucket or reference an existing one on
+a MongoDB database. Call the ``GridFSBuckets.create()`` helper method with
+a database instance as the parameter to instantiate a ``GridFSBucket``.
+You can use the ``GridFSBucket`` instance to call read and write operations on
+the files in your bucket.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start createGridFSBucket
+   :end-before: end createGridFSBucket
+
+.. _gridfs-create-custom-bucket:
+
+To create or reference a bucket with a custom name other than the default name
+``fs``, pass your bucket name as the second parameter to the ``create()``
+method as shown below:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start createCustomGridFSBucket
+   :end-before: end createCustomGridFSBucket
+
+For more information on the classes and methods mentioned in this section,
+see the following API documentation:
+
+- :java-docs:`GridFSBuckets <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBuckets.html>`
+- :java-docs:`GridFSBucket <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBucket.html>`
+
+.. _gridfs-store-files:
+
+Store Files
+-----------
+
+To store a file in a GridFS bucket, you can either upload it from an instance
+of ``InputStream`` or write its data to a ``GridFSUploadStream``.
+
+For either upload process, you can specify configuration information such
+as file chunk size and other field/value pairs to store as metadata. Set
+this information on an instance of ``GridFSUploadOptions`` as shown in the
+following code snippet:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadOptions
+   :end-before: end uploadOptions
+
+See the :java-docs:`GridFSUploadOptions <apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSUploadOptions.html>`
+API documentation for more information.
+
+Upload a File Using an Input Stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section shows you how to upload a file to a GridFS bucket using an input
+stream. The code example below shows how you can use a ``FileInputStream`` to
+read data from a file in your filesystem and upload it to GridFS by performing
+the following operations:
+
+- Read from the filesystem using a ``FileInputStream``.
+- Set the chunk size using ``GridFSUploadOptions``.
+- Set a custom metadata field called ``type`` to the value "zip archive".
+- Upload the file, specifying the file name as "myProject.zip".
```

Comment:
In the code snippet, I see the file path "/path/to/project.zip". Should this line instead read "project.zip"? If not, can we add a note or something explaining why the name changed? Unless I missed something.

Reply:
I have seen "/path/to" prefixed to describe the filesystem path in various MongoDB documentation without additional explanation. If this doesn't seem apparent, I can remove it.

Reply:
Oh ok, sorry if I misread. I'll clarify source and destination files don't have the same name and do not need to. I wouldn't want the reader to mistakenly think the filenames had to be the same.

Reply:
ahh it's actually my misunderstanding -- I didn't realize that "specifying the file name as" applied to the file as saved in GridFS (which I suppose should be obvious, since the step starts with "upload the file"). Personally I lay the blame at the feet of whoever decided that documents stored in GridFS are just called files instead of gFiles or something.